### ISSUE TITLE
Migration to coroutines

### ISSUE THREAD

**Issue by jraska** [jraska/github-client#603]
The community is with Coroutines already, though some problems need to be solved first for proper development

- [ ] Respecting the Coroutines context by Retrofit - things are currently executed on OkHttp thread pool
- [x] Idling in tests for reliable UI tests.
- [x] Lazy liveData
- [x] Testing by unit tests

**Comment by jraska** [comment]
Related to threading problems https://github.com/square/retrofit/issues/3513#issuecomment-789619816

The context is not propagated - regardless with which context we use the OkHttp thread pools is the on executing.
<img width="1020" alt="Screenshot 2021-09-13 at 21 37 49" src="https://user-images.githubusercontent.com/6277721/133145908-9927cf3d-f12c-496d-97c9-93a77a4671a9.png">
<img width="762" alt="Screenshot 2021-09-13 at 21 39 30" src="https://user-images.githubusercontent.com/6277721/133146048-35000f28-14e1-4cd8-9f8e-eb030f9fd330.png">

Yields: `Network thread: OkHttp`

It works when used:
<img width="502" alt="Screenshot 2021-09-13 at 21 52 10" src="https://user-images.githubusercontent.com/6277721/133147621-4a4f6bc9-0f71-4312-8ac1-d8b761d88109.png">
<img width="859" alt="Screenshot 2021-09-13 at 21 51 55" src="https://user-images.githubusercontent.com/6277721/133147644-d10515d7-137f-41f7-9872-3bba42781e2f.png">

But this looks ugly and wrong, however works and yields: `Network thread DefaultDispatcher-worker-3`

**Comment by jraska** [comment]
Anyway first the UI testing will be tried to see if it will be a problem - since likely most of Android engineers live with OkHttp thread - I might as well.

**Comment by jraska** [comment]
Lazy livedata solved by `asLiveData`

**Comment by lordcodes** [comment]
I think using their own thread is deliberate, as in Coroutines the contract is that a suspend function should be callable from the main thread.
In general with coroutines you don't need to specify the context dispatcher as often as you did with RxJava, as libraries like DB (Room) and Network (Retrofit/OkHttp) with Coroutines support handle the correct thread for their own operations.

**Comment by jraska** [comment]
Thanks for the input :) 

The thing is if it is deliberate I don't like it - it feels like we are getting back to early Android days, where each library - OkHttp in this case would spin it's own thread pool for async operations. I'm fine if this option is here, I just don't like it being implicit without opt out, because besides not respecting the threads used - as RxJava adapters do - it can bring issues with error handling like for example [here](https://github.com/square/retrofit/issues/3513#issuecomment-789619816) + I observed some other cases.


Btw overall whilst finally diving into the coroutines my impressions is that by evolution it got almost the same problems as RxJava plus different ways of doing the same thing - so for me the only difference of RxJava and Coroutines is now the hype :D which at the ends rules the Android world.

**Comment by lordcodes** [comment]
I much prefer Coroutines and Flow to RxJava, way simpler to work with and causes much less mental load I find. However, I will agree I have found the dispatcher/threading part harder to get my head around.

**Comment by jraska** [comment]
Yes, engineers tend to like coroutines as "easier". However the production incidents I've seen with them quite well copy the issues seen with RxJava - e.g. improper error handling. Obviously all of them were incorrect usage of coroutines, not a problem within coroutines, but that applies to RxJava as well.

One thing with Coroutines which particularly bothers me is there is still not proper support to Idling resources - after several years for [this issue to be open](https://github.com/Kotlin/kotlinx.coroutines/issues/242) - there is a loot of workaround options with different degree of problems. This was particularly main thing why I was waiting so long - I don't find something production ready if it doesn't work reliably with UI tests. Now I will just pick some of the workarounds. 🤷

**Comment by jraska** [comment]
Hey @lordcodes - I'm hitting the problem with not propagating context now in unit tests so I would be curious on your thoughts.

I have a [RepoDetailViewModel](https://github.com/jraska/github-client/blob/master/feature/repo/src/test/java/com/jraska/github/client/repo/RepoDetailViewModelTest.kt) test, which checks the whole model with fully setup DI only with mocked network.

With Rx the schedulers are properly propagated and in tests are replaces for `trampoline`, therefore the network requests run also synchronously and the test works.
With the coroutines and Retrofit not propagating it, the OkHttp spins off a separate thread even if [the context of the coroutines is `Dispatchers.Unconfined`](https://github.com/jraska/github-client/commit/c92fb1a533c64ce349d86d05b3a0e1b03337d7b5#diff-67feea41760feff826caf547385d504530269fe3139982f748c37f60269e2659R30). The result is the network requests ae not blocking and the test fails on asserting too early, getting `Loading` state instead of the loaded one.

If we us the `Call<Dto>` as a return type and call `execute.body()!!` the network request is executed as blocking within coroutine (which goes against coroutines as principle), but then the proper context is being kept and code becomes testable. See this commit moving from working solution to the pretty - not woking one. https://github.com/jraska/github-client/commit/f5e2e0518f8f46b5a3ca89c46a5e0e621537bb51

The questions is: **How to solve this nicely?** 
By now I'm very tempted to stick of using `Call<Dto>` with some extension function to make it less ugly, but for proper testing with real netwok calls I fail to see aa different solution given Retrofit coroutines ignore passed context - dispatcher.

Whole changeset can be seen here: https://github.com/jraska/github-client/compare/repo-detail-to-coroutines?expand=1

**Comment by lordcodes** [comment]
`runBlocking` will block whilst waiting for result, so test function shouldn't end. Therefore I assume it means the 'Flow' is completing early. As within runBlocking you are collecting all the values from the Flow.

**Comment by jraska** [comment]
Yeah so the `runBlocking` didn't work for me. It does work for example [here](https://github.com/jraska/github-client/blob/repo-detail-to-coroutines/feature/repo/src/test/java/com/jraska/github/client/repo/model/GitHubApiRepoRepositoryTest.kt#L20).

But it feels like it doesn't work when the whole new context is used like in [here](https://github.com/jraska/github-client/blob/repo-detail-to-coroutines/feature/repo/src/main/java/com/jraska/github/client/repo/RepoDetailViewModel.kt#L45) or when the OkHttp launches it's own thread 🤷

**Comment by lordcodes** [comment]
AFAIK changing context has no impact on runBlocking. RunBlocking is simply that it will launch a Coroutine, then wait until it completes before returning execution. What you described, is that it asserted earlier. The assert is inside runBlocking so it means the code before the assert completed before you thought it should.

runBlocking doesn't alter the execution of the coroutine, so that each line inside runBlocking will block. It simply effects any code you have after the runBlocking call, i.e. in this case blocking the test body until the coroutine launched inside runBlocking completes.

Compared to 'launch' which will launch coroutine and then return, so test would end immediately.

**Comment by jraska** [comment]
Yeah anyway the problem still stays :D

If I use `suspend` function within coroutine, passing parameter of synchronous dispatcher doesn't make the flow to complete synchronously, because it spins extra OkHttp thread instead so I don't see a way how to test such logic using retrofit network request with suspend call. 

`RepoDetailViewModelTest` -> `RepoDetailViewModel` -> `RepoRepository` -> `RepoGitHubApi` using Retrofit.

**Comment by lordcodes** [comment]
If you launch a Coroutine and block until it receives a result, it doesn't matter what Dispatcher or threads are used within that suspend function within the Coroutine. Suspend functions return a result, they can't return multiple results or no result. I.e. It's about making whatever code calls the suspend function wait until the result.

If you do 'runBlockingTest { retrofitApi.request() }', where request() is suspend. runBlockingTest will not complete until the request() completes, regardless of Dispatchers or threading within Retrofit/OkHttp.

Also, within tests there is 'runBlockingTest' within the Test Coroutines package, I don't know exactly how it differs to regular runBlocking, worth looking at that.

There is a key different in Coroutines/Flow vs RxJava. RxJava subscribe function completes immediately. That is not the case with collecting Flows or running suspend functions within a Corotuine. Although is the case with the actual 'launch' function.

I can't really tell which test isn't working as RepoDetailViewModelTest isn't in that changeset you linked to above.

**Comment by jraska** [comment]
This commit should hopefully explain it. :) https://github.com/jraska/github-client/commit/7479d7d938695ca402dbd671ec9667417acecae5

> Also, within tests there is 'runBlockingTest' within the Test Coroutines package, I don't know exactly how it differs to regular runBlocking, worth looking at that.

In case the `Dispatchers.Unconfined` the flow behaves like Rx. If `RxJavaCallAdapter` for Retrofit would do the same as coroutines - not respecting the passed `Scheduler` and silently doing `subscribeOn(sometingNew)` the RxJava would have the same problem here :) 

Anyway the length of this discussion shows me why Coroutines are no simpler than Rx :D

**Comment by jraska** [comment]
The real problem I try to solve: How to test fully injected ViewModel only with LiveData where the only mocked part in tests is network? :D

**Comment by lordcodes** [comment]
> This commit should hopefully explain it. :) [7479d7d](https://github.com/jraska/github-client/commit/7479d7d938695ca402dbd671ec9667417acecae5)

The point I was trying to make, it is less about the code here and more about the test around it.

> The real problem I try to solve: How to test fully injected ViewModel only by LiveData where the only mocked part in tests is network? :D

This highlights an issue. 'LiveData' is hot, so you need to be sure the Coroutines -> LiveData conversion is blocking correctly.

**Comment by jraska** [comment]
Yes, but I rely on the `asLiveData` from ktx, which I think I should? 
I think it blocks correctly :D until the point when it reaches Retrofit and OkHttp :D

**Comment by lordcodes** [comment]
I disagree that Retrofit and OkHttp threading causes a Flow that you created around a suspend function, causes it not to block.

I would need to try breaking it down and testing each part to work out where the problem is, my gut says it's in the 'Flow' part or the conversion to LiveData.

* Retrofit suspend called straight from test, with runBlockingTest
* ViewModel Flow logic collected straight from test, with runBlockingTest

**Comment by jraska** [comment]
Will try :) will help with learnings this ~hyped~amazing technology :D

Meanwhile [this dirty PoC](https://github.com/jraska/github-client/commit/3cabd9c1fbfae3d0cf8a433152fe9ac24a5f889c#diff-d59e5c144923a2e996fd3f838a7ea9d0f5d716c4cc25c6193a11e01726cbc4beR37) shows how to reliably idle in UI tests. Passed correctly 30+ times so I start to have trust in it :)

**Comment by lordcodes** [comment]
> Meanwhile [this dirty PoC](https://github.com/jraska/github-client/commit/3cabd9c1fbfae3d0cf8a433152fe9ac24a5f889c#diff-d59e5c144923a2e996fd3f838a7ea9d0f5d716c4cc25c6193a11e01726cbc4beR37) shows how to reliably idle in UI tests. Passed correctly 30+ times so I start to have trust in it :)

Okay, will try it out, thanks.

**Comment by jraska** [comment]
[This `IdlingDispatcher`](https://github.com/jraska/github-client/blob/repo-detail-to-coroutines/core-android-testing/src/main/java/com/jraska/github/client/android/test/IdlingDispatcher.kt) is production(test) ready.

If there isn't a library already maybe I can create one :D

**Comment by jraska** [comment]
So I think we are somehow both right with our previous statements :D 

> - Retrofit suspend called straight from test, with runBlockingTest
> - ViewModel Flow logic collected straight from test, with runBlockingTest

Both of these cases work correctly so the issue is triggered by using `asLiveData`, but I still think there is a problem in OkHttp caused by the thread switch - I'm still trying to understand if it is a feature or a problem :D 

- In case we directly execute coroutines the runBlocking makes sure everything completes - makes perfect sense
- If things go through LiveData, then it depends on the Dispatcher the tests using LiveData [here](https://github.com/jraska/github-client/blob/repo-detail-to-coroutines/feature/repo/src/test/java/com/jraska/github/client/repo/RepoDetailViewModelTest.kt#L43) pass if the IO dispatcher passed is `Dispatchers.Unconfined`, but fail if the real `Dispatchers.IO` is used. This is the behaviour I want.

The thing is if you add inside the flow for example
```
withContext(Executors.newSingleThreadExecutor().asCoroutineDispatcher()) { longSuspendFun() }
```
like [here](https://github.com/jraska/github-client/blob/repo-detail-to-coroutines/feature/repo/src/main/java/com/jraska/github/client/repo/model/GitHubApiRepoRepository.kt#L18), tests will start to fail as the Flow is no longer synchronous, because the `Dispatchers.Unconfined` is not used and new Dispatcher is used. Retrofit with OkHttp behave exactly the same as if I use the withContext(Executors.newSingleThreadExecutor().asCoroutineDispatcher()) 

So maybe updated definition of a problem: 
How to parametrise coroutine code, which is triggered by not coroutine code a way that it is synchronous? in case we don't call coroutine directly in tests. 
- The `runBlocking` does not work, but this is a feature  :D 
- The `Dispatchers.Unconfined`, but only unless someone changes context/thread within some of the coroutines.


I think now might be just easir to have a call with live coding :D

**Comment by jraska** [comment]
# Some comparison after rewriting from RxJava to Coroutines

### Line numbers 
comparison of the 3 PRs:
+188 −50, +124 −117, +70 −155 = **+60 lines of code** using Coroutines - If we take in account no support of Idling for Coroutines, we could count -56 lines of `IdlingDispatcher` and + 10 lines of RxIdler integration we might be at **+14 lines** 

### Binary size 
After obfuscating in Play the Bundle difference is **+0.05 MB** with Coroutines compared to RxJava

### Dependencies
-RxJava, - RxAndroid, - RetrofitRxAdapter, - RxIdler
+Coroutines JVM, Coroutines Android, Coroutines Test, lifecycle-viewmodel-ktx, lifecycle-livedata-ktx

What I find positive here is that the author and maintainer is now JetBrains/Google compared to individuals before

### Threads used
Before only RxSchedulers thread pools, now also library thread pools like OkHttp - I don't want that, but this is what Coroutines promote

### Testing support
- RxJava has unit testing embedded in, Idlers for UI testing ready - RxIdler
- Coroutines test still has majority of api as experimental, no support for idling.
- Overall Coroutines feel like there is many different test practices which brings confusion.

### Public API used
- Rx: `Observable`, `Completable`, `Single`, `PublishSubject`, `Scheduler`, `Schedulers`, `AndroidSchedulers`, `RxJavaPlugins`, `Disposable`, `CompositeDisposable`, `BackpressureStrategy`, `Consumer`, `Action`
- Coroutines: `suspend`, `Dispatchers`, `CoroutineDispatcher`, `CoroutinesContext`,`Flow`, `flow`, `MutableSharedFlow`, `FlowCollector`, `Job`, `async`, `awaitAll`, `withContext`, `Deferred`, `InternalCoroutinesApi`, `DelicateCoroutinesApi`, `BufferOverflow`, `runBlocking`

It is not trivial here to compare apples with apples as RxJava has typos and Coroutines has bunch of extension functions - I might try to dive a bit more into number of methods used etc.

### Maturity - moving to more subjective line
- RxJava is fully mature and the major version 3.0 speaks for it - RxJava had many problems in version 1.x and I would claim Coroutines really need version 2.0
- Half of the Coroutine APIs are deprecated or experimental or marked with some other fancy annotation - I've never seen a tool having so much random annotation hints.
- Coroutines are ambiguous as they evolved and it still feels like a teenager discovering its body, hence I feel they need 2.0 to choose on strategy for one thing, finalise testing guidance instead of deprecating half of the API and annotating each method with something.
- What is also mind-blowing to me is that too many people consider launching code in constructor as ok, even [official docs.](https://developer.android.com/topic/libraries/architecture/coroutines#lifecyclescope)

### Traceability & Debugging
RxJava requires a bit of a mindset shift, but there is zero synthetics - you can always see the code exactly as it was written whilst navigating. Though Coroutines do compile time magic appending `Continuation`, so the code just does not exist. Also even if RxJava stacks can be mind-blowing, it makes sense as it contains names of the operator used, so you get ida what is happening. Both RxJava and Coroutines us million lambda so the generated classes/methods are comparable.

### Support
- Here Coroutines is clear winner - Google + JetBrains push it massively 
- The language and tooling integration of Coroutines is as good as it can be given the magic Coroutines do :D 
- Depending on Coroutines is way safer than RxJava - only community individuals maintain it. If David Karnok steps away, RxJava will likely lose quality.


### TLDR:
Coroutines without a giant support and hype they receive would hardly be considered better than RxJava. However the support and developers sentiment plays a huge role and considering the tradeoffs I will stick to Coroutines now as "the future". Though i still strongly stand against the claims they are just simpler - they were, before they had to handle all use cases Rx did :D
